hw/misc/hermes.c

```diff
@@ -458,6 +475,108 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
                 addr);
 }
 
+static inline void mark_exec_finished(HermesState *hermes, int engine)
+{
+    bar0_init.cmdctrl[engine].ehcmddone = HERMES_CMD_FINISHED;
+    trace_hermes_msix_notify(H2C_CHANNELS + C2H_CHANNELS + engine);
+    msix_notify(PCI_DEVICE(hermes), H2C_CHANNELS + C2H_CHANNELS + engine);
+}
+
+static void hermes_exec(HermesState *hermes, int engine)
+{
+    struct hermes_cmd *command = &bar0_init.commands[engine];
+
+    void *bar4_base = memory_region_get_ram_ptr(&hermes->bar4_mem_reg);
+
+    char *errmsg;
+    int32_t rv;
+    uint64_t ret;
+    bool elf;
+
+    unsigned char prog_slot = command->req.run_prog.prog_slot;
+    unsigned char data_slot = command->req.run_prog.data_slot;
+
+    void *program_base = bar4_base + bar0_init.ehpsoff +
+                         (prog_slot * bar0_init.ehpssze);
+    void *data_base = bar4_base + bar0_init.ehdsoff +
+                      (data_slot * bar0_init.ehdssze);
+
+    command->res.cid = command->req.cid;
+
+    if (prog_slot >= bar0_init.ehpslot) {
+        command->res.status = HERMES_STATUS_INVALID_PROG_SLOT;
+        goto FINISH_EXEC;
+    }
+
+    if (data_slot >= bar0_init.ehdslot) {
+        command->res.status = HERMES_STATUS_INVALID_DATA_SLOT;
+        goto FINISH_EXEC;
+    }
+
+    if (command->req.opcode != HERMES_OPCODE_EXECUTE_SLOT) {
+        command->res.status = HERMES_STATUS_INVALID_OPCODE;
+        hermes_bar_warn_invalid_command(command->req.opcode, engine);
+        goto FINISH_EXEC;
+    }
+
+    /* This is a hack to load a new program into a dirty VM!
+     *  https://github.com/iovisor/ubpf/pull/63
+     *  https://github.com/Eideticom/eid-hermes-qemu/pull/14#discussion_r622587139
+     */
+    if (hermes->ubpf[engine].engine->insts) {
+        free(hermes->ubpf[engine].engine->insts);
+        hermes->ubpf[engine].engine->insts = NULL;
+    }
+
+    /* Check magic number (first 4 bytes), to see if program is in ELF format */
+    elf = bar0_init.ehpssze >= SELFMAG && !memcmp(program_base, ELFMAG, 4);
+    if (elf) {
+        rv = ubpf_load_elf(hermes->ubpf[engine].engine, program_base,
+                           command->req.run_prog.length, &errmsg);
+    } else {
+        rv = ubpf_load(hermes->ubpf[engine].engine, program_base,
+                       command->req.run_prog.length, &errmsg);
+    }
+
+    if (rv < 0) {
+        warn_report("%s", errmsg);
+        command->res.status = HERMES_STATUS_EBPF_ERROR;
+        command->res.run_prog.ebpf_ret = rv;
+        goto FINISH_EXEC;
+    }
+
+    ret = ubpf_exec(hermes->ubpf[engine].engine, data_base, bar0_init.ehdssze);
+
+    if (ret == UINT64_MAX) {
+            command->res.status = HERMES_STATUS_EBPF_ERROR;
+    }
+
+    command->res.status = HERMES_STATUS_SUCCESS;
```

Comment:
this should be in an else, otherwise we overwrite the status you just set above in the error case